python/pyspark/sql/column.py

```diff
@@ -406,7 +406,13 @@ def substr(self, startPos, length):
         [Row(col=u'Ali'), Row(col=u'Bob')]
         """
         if type(startPos) != type(length):
-            raise TypeError("Can not mix the type")
+            raise TypeError(
+                "startPos and length must be the same type. "
+                "Got {startPos_t} and {length_t}, respectively."
```

Comment:
If PySpark always needs to check the types, are we doing the same things in all the other function calls?

In addition, why not directly checking 
```Python
if isinstance(length, (int, long)):
```